### DISCUSSION TITLE
Uploading files that can be read by python code in a notebook

### DISCUSSION THREAD

**Discussion by matthewcmead** [jupyterlite/jupyterlite#91] [Q&A] [answered]
I'm hoping to be able to upload files to the jupyterlite instance running in the browser, and then see those files with python code (as in f = open('/home/web_user/file...', 'r') or similar).  Unfortunately I am unable to find anywhere in the pyodide runtime's view of the filesystem that corresponds to the files managed in jupyterlite's jupyterlab instance.

Is this doable?  If not, where can I plugin to assist in making this a reality?

Thanks!

**Comment by jtpio** [comment]
Thanks @matthewcmead.

For now notebooks and files are saved in the browser local storage (`IndexedDB` or `localStorage` depending on availability). These files are not yet "mounted" into the kernel.

Ideally we should indeed be able to see these files from the kernels. It's not clear yet how this should work, and whether it will be possible to expose these files to all kernels, or to Pyodide only.

In the meantime, you might be able to work around that by placing files is a specific folder, and requesting them from Python with something like this:

```python
from js import fetch

# example retrieving the default html file on the demo website
res = await fetch('https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html')
text = await res.text()

filename = 'myfile.html'

with open(filename, 'w') as f:
    f.write(text)

with open(filename) as f:
    data = f.read()
    
print(data)
```

**Reply by psychemedia** [reply]
That errors for me:

```
Traceback (most recent call last):
  File "<console>", line 4, in <module>
  File "/lib/python3.8/asyncio/futures.py", line 260, in __await__
    yield self  # This tells Task to wait for completion.
  File "/lib/python3.8/asyncio/tasks.py", line 349, in __wakeup
    future.result()
  File "/lib/python3.8/asyncio/futures.py", line 178, in result
    raise self._exception
TypeError: exceptions must derive from BaseException
```

**Reply by psychemedia** [reply]
Hmm...maybe it didn't like the MIMEtype of file type I was trying to request...

**Comment by bollwyvl** [comment]
Yeah this is possible. Similarly, if the "http" server was visible inside
kernels, that could be very fun.

Pretty much all the client kernels have some way to "shell out" to the
browser js environment. So, as a band-aid in pyolite, we could add some
magics that push/pop localstorage, indexdb, or whatever you might be
backing your contents.

It might even work against the in-browser contents manager right now over
"http" (by using get()/put()) but something more portable would be
preferred. I'm not sure if there is a functioning async http client that
installs well with micropip. tornado has no dependencies, so might be a
good experiment.

A more advanced form would be to actually provide a mount to the wasm file
system, but I don't know much about such a beast, and wouldn't know where
to start. different kernels that are both deployed with wasm might have
different approaches to storage, and some might have split wasm/worker
implementations.

**Comment by matthewcmead** [comment]
Thanks for all the replies.  What I'm hoping to do is allow for arbitrary files to be dragged into the environment once it is up and running, operated on from within python, written into the file store, and downloaded.  This would allow for various "toolbox" type notebooks to be built into jupyterlite that can perform tasks on user data (when enabled with upload/download).

I was able to stitch together enough to read files out of the js context's indexedDB, but it would be a lot cleaner if localforage were exposed to the pyodide kernel in a manner that the files store could be easily manipulated.  How hard would this be to accomplish?

I agree the long-term goal of providing a vfs layer / mount to the kernels would be best, but perhaps an intermediate step that allows writing of a custom jupyter storage manipulation (a la an object store like S3) library would work?

Thanks again!

**Reply by bollwyvl** [reply]
> enabled with upload/download

Looks like download is coming on #89. Is upload misbehaving for you?

> localforage were exposed to the pyodide kernel 

I'm _pretty_ sure pyolite is running inside a WebWorker (@jtpio would have to confirm) without access to much, but presumably it _could_ have access to the same localStorage/IndexDB, but currently doesn't (vs the [pyodide console](https://pyodide.org/en/stable/console.html). Man, we need some of that TTY goodness!

> was able to stitch together enough to read files out of the js context's indexedDB

Got anything to share?

> vfs layer / mount to the kernel

It looks like there is some [good progress](https://github.com/pyodide/pyodide/issues/613) with [BrowserFS](https://github.com/jvilk/BrowserFS) which already plugs into the emscripten filesystem. This might be The Way, though sadly it looks semi-abandoned. Having `/home/jovyan` or something so it acts like binder, as it's the most canonical thing we have, with an accurate `Path.cwd`, etc. would make it seem much more "real".

>  custom jupyter storage manipulation

yeah, I don't anticipate us touching any s3 for some time, first-party, but eventually there will be a toolset, either derived from (or exactly) `jupyter labextension build` that will allow for hacking more things into the serverlite `ContentsManager` and whatever we're able to bridge into exotic filesystems like pyolite's magic in-memory system.

**Reply by jtpio** [reply]
> I'm _pretty_ sure pyolite is running inside a WebWorker

Yes that's right: 

https://github.com/jtpio/jupyterlite/blob/7dd0ae7b95e8dd896ec8a9ac6869bd4f8bdf49fc/packages/pyolite-kernel/src/kernel.ts#L28-L36

The main reason for this for now was to prevent blocking the main thread, and better isolate multiple kernels running on the same page.

> but perhaps an intermediate step that allows writing of a custom jupyter storage manipulation (a la an object store like S3) library would work?

This is definitely something that should (will) be supported at some point soon when there is tooling to develop serverlite extensions as federated extensions. Then it should be possible to either replace or extend the current `ContentsManager` with something else.

**Reply by matthewcmead** [reply]
Thanks again for the thoughtful replies on this topic.

Download and Upload from the browser user to jupyterlite is working fine.

Sorry, I didn't mean to suggest literally be using S3.  Rather, I was thinking of having a pyodide-accessible interface to localforage _files storage which operated like an object store.  I.e., get/put for the most part.  That way a python notebook could manage moving files between the in-ram filesystem and the jupyterlab storage.

I do think for a long-term solution having the jupyterlab storage exist somewhere in the ram file tree as a virtual filesystem would be preferable -- it would be transparent and operate with standard python I/O like most users are accustomed to.

I've not done much typescript/javascript work, so I don't know what all I can offer in way of help without a learning curve, but I'm happy to put in some effort on this if someone has pointers on stubbing access to the _files storage in the pyodide javascript context.

Thanks again!

**Comment by bollwyvl** [comment]
Yet another angle, that may require some upstream assistance, is that pyodide could be built to read/write directly to IndexedDB instead of memory:

https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs

If this were the case (or available from a specific build) then we could likely do a custom localForage driver that quacked the same way, and all kernels that were similarly compiled would have access to the same files, which would be quite powerful indeed.

**Reply by bollwyvl** [reply]
Some progress here:
https://github.com/pyodide/pyodide/pull/1596

if we get that one home, and a release is cut, from the js side, right after pyodide starts:
```js
pyodide.mkdir("/home/jo")
pyodide.mount("/home/jo")
await pyodide.syncfs(true)
```
and now those files will be visible to pyodide. We can fiddle with the setup so that pyodide starts in there, and all your files would be hot, as of when it synced, whether they came from the `localForage` layer, or from the server... but we want to be careful, as one demo video (which will be very cool when easy to add) might gum up the works tremendously.

It's not magic, though, and as writing to indexdb isn't free, that `syncfs` has to be called both times, but there would be runtime access to:

```py
import js
await js.pyodide.syncfs(False)
```

...which would push everything back to indexeddb. pyolite in turn could wrap that with magics, etc. Further, when the UI says _Save_, that could trigger a `syncfs`, which probably jives with what people would expect, but you wouldn't get, e.g. streaming logs out, etc.

**Reply by bollwyvl** [reply]
Good news/bad news:

- good: the full emscripten API is available from a private member on pyodide, e.g. `js.pyodide._module`
- bad: the various backends still need to be compiled in before we could do anything with it

**Reply by bollwyvl** [reply]
less bad:
```py
>>> import js
>>> js.Object.keys(js.pyodide._module.FS)
root,mounts,devices,streams,nextInode,nameTable,currentPath,initialized,ignorePermissions,trackingDelegate,tracking,ErrnoError,genericErrors,filesystems,syncFSRequests,lookupPath,getPath,hashName,hashAddNode,hashRemoveNode,lookupNode,createNode,destroyNode,isRoot,isMountpoint,isFile,isDir,isLink,isChrdev,isBlkdev,isFIFO,isSocket,flagModes,modeStringToFlags,flagsToPermissionString,nodePermissions,mayLookup,mayCreate,mayDelete,mayOpen,MAX_OPEN_FDS,nextfd,getStream,createStream,closeStream,chrdev_stream_ops,major,minor,makedev,registerDevice,getDevice,getMounts,syncfs,mount,unmount,lookup,mknod,create,mkdir,mkdirTree,mkdev,symlink,rename,rmdir,readdir,unlink,readlink,stat,lstat,chmod,lchmod,fchmod,chown,lchown,fchown,truncate,ftruncate,utime,open,close,isClosed,llseek,read,write,allocate,mmap,msync,munmap,ioctl,readFile,writeFile,cwd,chdir,createDefaultDirectories,createDefaultDevices,createSpecialDirectories,createStandardStreams,ensureErrnoError,staticInit,init,quit,getMode,findObject,analyzePath,createPath,createFile,createDataFile,createDevice,forceLoadFile,createLazyFile,createPreloadedFile,indexedDB,DB_NAME,DB_VERSION,DB_STORE_NAME,saveFilesToDB,loadFilesFromDB,FSNode,FSStream
```

So there is already a lot to play with.

**Comment by bollwyvl** [comment]
Update here:
- https://github.com/pyodide/pyodide/pull/1596 is still in-progress, but has been updated to include all the backends, as the latest pyodide now supports `nodejs`
  - in support of that, we've gotten https://github.com/pyodide/pyodide/pull/1692 merged, which would provide a "blessed" way to access the `FS` module

Once these are in and start hitting the nightlies, https://github.com/jupyterlite/jupyterlite/pull/178  could be updated to start using these APIs. In the near term, we could start caching wheels, etc. in indexedDB, which would make a second kernel start up _significantly_ snappier (even in private browsing for browsers that support indexeddb there).

User python code would also then have the ability to create folders/files in the filesystem, but wouldn't (by design) have a handle on the outer contents manager... and as this isn't magically going to become free to keep them in sync, even with `indexedDb`.

An approach we could _already_ have might be to give `pyolite` the ability to `GET` from/`PUT` to the contents manager API, so code might look like:

```py
import pyolite
# while it's thinking about that, upload something
await pyolite.contents.get("the-thing-i-uploaded.txt", "a-local-name.txt")  # comes into emscripten
# do something to the notebook
await pyolite.contents.put("a-local-name.txt", "the-thing-i-uploaded.txt")           # goes back into the browser
```

where we hide all the marshalling in `pyolite`.

**Comment by konwiddak** [comment]
I've managed to read data from the indexedDB inside a notebook. I don't have the know how to incorporate this into a single function since it only seems to work when executed in separate cells - something needs to allow the transaction to complete. I'm sure there's a simple solution to that. I also presume this could be made to work with saving using a put instead of a get.

Cell 0:

```
import js
from io import StringIO
import pandas as pd
```

Cell 1:

```
op = js.self.indexedDB.open("JupyterLite Storage")
```

Cell2:

`idbdb = op.result`

Cell 3:

```
tr = idbdb.transaction("files")
obs = tr.objectStore("files")
req = obs.get("test_data.csv", "key")
```

Cell 4:
```
csv = req.result.content
csv_io = StringIO(csv)
pd.read_csv(csv_io, sep = "\t")
```

![jupyterlite_read_csv](https://user-images.githubusercontent.com/48279914/128383409-77fa53e8-2f20-4c8d-8a38-1aec84018c6d.png)

**Reply by bollwyvl** [reply]
Rad hack :muscle: !

Digging into what those various things _are_ with some debugging and MDN, I was able to get this _further_ hack to work in one `async` function:

```py
async def get_contents(path):
    from asyncio, js
    q = asyncio.Queue()
    db = js.self.indexedDB.open("JupyterLite Storage")
    db.onsuccess = q.put_nowait
    await q.get()
    tr = idbdb.transaction("files")
    obs = tr.objectStore("files")
    req = obs.get(path, "key")
    req.onsuccess = q.put_nowait
    await q.get()
    return req.result.to_py()
```

Which would then be called, _somewhat_ awkwardly with:

```py
import pandas as pd
from io import StringIO
pd.read_csv(StringIO((await get_contents("iris.csv"))["content"]), sep = "\t")
```

That is... not awful! I mean, this is what a regular `ipykernel` would have to do if its storage and the `ContentsManager` were not trivially co-located, and it had to talk to the REST API, but no token discovery hacks are needed, which is nice. Of note, this _didn't_ work with the `iris.csv` we ship in the demo: i had to open it and re-save it to get it hot, but pyodide 0.18 also opens the door for more use of bidrectional `comlink` things, so we might be able to wrap _whatever_ we're backing it with, which would be clutch for e.g. big images, databases, etc.

If there ends up being no way on (a follow-up to) #274 to transparently trick emscripten into doing the work for us (e.g. redoing our IndexedDB layout to be emscripten-compatible), we should definitely consider adding this (and similar `put_contents`) to `pyolite`. Or another approach would be to implement it as something like `pathlite.Path`, so that the internal structure of the Contents model is a little less front-and-center (but it would be a shame to just drop it on the floor). Probably would still need `await` though... so maybe not so much gained.

Thanks again :heart: !

**Reply by bollwyvl** [reply]
gah: broke something in there... will fix it when i get a chance!

**Reply by bollwyvl** [reply]
Here's a cleaned-up version:

```py
import js, asyncio

async def get_contents(path):
    """use the IndexedDB API to acess JupyterLite's in-browser (for now) storage
    
    for documentation purposes, the full names of the JS API objects are used.
    
    see https://developer.mozilla.org/en-US/docs/Web/API/IDBRequest
    """
    # we only ever expect one result, either an error _or_ success
    queue = asyncio.Queue(1)
    
    IDBOpenDBRequest = js.self.indexedDB.open(DB_NAME)
    IDBOpenDBRequest.onsuccess = IDBOpenDBRequest.onerror = queue.put_nowait
    
    await queue.get()
    
    if IDBOpenDBRequest.result is None:
        return None
        
    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readonly")
    IDBObjectStore = IDBTransaction.objectStore("files")
    IDBRequest = IDBObjectStore.get(path, "key")
    IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait
    
    await queue.get()
    
    return IDBRequest.result.to_py() if IDBRequest.result else None
```

Full gist:
https://gist.github.com/bollwyvl/132aaff5cdb2c35ee1f75aed83e87eeb

**Reply by psychemedia** [reply]
Yeah, getting there... :-)

I note that you don't need to edit the CSV file, simply double clicking the `iris.csv` file to open it and then saving it works too. (Actually, I didn't check what happens if you just double click to open the CSV file in the  and then try to load it via `get_contents()`?)

For simple edu purposes, this is not so bad because a good way to get novices to work through the material would be to get them to preview the contents of things like a CSV file first. In that context, the workflow (even though it is currently a hack to make things work) could seem to be almost quite natural...

**Reply by konwiddak** [reply]
Forgive me for the sin I am about to commit!

@bollwyvl, if we take your code, return a StringIO and override the "open" builtin then we're not far off regular python. I'm not sure it's possible to remove the "await" in front of the "open", which is a shame because that would make this a pretty elegant hack. (If anyone has any ideas?)

```
import js, asyncio
import micropip
from io import StringIO
from builtins import open
import builtins
import pandas as pd

DB_NAME = "JupyterLite Storage"


async def get_contents(path, *args):
    """use the IndexedDB API to acess JupyterLite's in-browser (for now) storage

    for documentation purposes, the full names of the JS API objects are used.

    see https://developer.mozilla.org/en-US/docs/Web/API/IDBRequest
    """
    # we only ever expect one result, either an error _or_ success
    queue = asyncio.Queue(1)

    IDBOpenDBRequest = js.self.indexedDB.open(DB_NAME)
    IDBOpenDBRequest.onsuccess = IDBOpenDBRequest.onerror = queue.put_nowait

    await queue.get()

    if IDBOpenDBRequest.result is None:
        return None

    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readonly")
    IDBObjectStore = IDBTransaction.objectStore("files")
    IDBRequest = IDBObjectStore.get(path, "key")
    IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait

    await queue.get()

    data = IDBRequest.result.to_py() if IDBRequest.result else None
    return StringIO(data['content'])
    
# Override builtin open.....
open = get_contents

# Almost like a normal file.... In theory I think writing could also be supported this way.
with await open('test_data.csv','r') as f:
    print(f.readline())
    
# Reads into pandas nicely
with await open('test_data.csv','r') as f:
    df = pd.read_csv(f)

df
```

**Reply by bollwyvl** [reply]
there are no sins in the browser, only things that work with fewer hacks. :imp: But yeah, i'd go with a `pathlite.Path` over hacking built-ins: this creates _real_ problems for things that use introspection internally.

the way things are, we don't have a good way to have library-level stuff that does `loop.run_until_completed` (see pip discussions), so folk are going to have to make peace with more asynchrony than they are perhaps used to in the meantime... having _anything_ that only needs _one_, built-in concurrency pattern (vs the _n_ the underlying JS often needs) is going to be much easier to document.

**Reply by psychemedia** [reply]
@bollwyvl This approach no longer seems to work for me in `jupyterlite==0.1.0a16` ?

**Comment by oscar6echo** [comment]
I tried a reasonable symmetric `put_contents()` function:

````py
async def put_contents(content, path, overwrite=False):
    """
    """
    # count existing
    queue = asyncio.Queue(1)
    
    IDBOpenDBRequest = js.self.indexedDB.open(DB_NAME)
    IDBOpenDBRequest.onsuccess = IDBOpenDBRequest.onerror = queue.put_nowait
    await queue.get()
    
    if IDBOpenDBRequest.result is None:
        return None
        
    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readonly")
    IDBObjectStore = IDBTransaction.objectStore("files")
    IDBRequest = IDBObjectStore.count(path)
    
    IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait
    await queue.get()
    
    count = IDBRequest.result
    print(f'count = {count}')
    
    if count == 1 and not overwrite:
        print(f'file {path} exists - will not overwrite')
        return 
    
    # add file
    value = {
        'name': os.path.basename(path), 
        'path': path,
        'format': 'text',
        'last_modified': dt.datetime.now().isoformat(),
        'content': content,
        'mimetype': 'text/plain',
        'type': 'file',
        'writable': True,
    }
    print(value)

    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readwrite")
    IDBObjectStore = IDBTransaction.objectStore("files")
    IDBRequest = IDBObjectStore.add(value, path)
    
    IDBRequest.oncomplete = IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait
    await queue.get()
    
    return IDBRequest.result.to_py() if IDBRequest.result else None
````

But when I call `await put_contents('azerty', 'data/test_3.csv', overwrite=False)` it breaks on:
+ line `IDBRequest = IDBObjectStore.add(value, path)`  
+ error: `JsException: DataCloneError: Failed to execute 'add' on 'IDBObjectStore': [object Object] could not be cloned.
`

Pls advise.

The objective is to make the saved file available for download, which jlab offers.


PS: This Jupyterlite endeavour looks really promising. Congratulations :clap:

**Comment by oscar6echo** [comment]
Here is a complete working solution to read and write a file which does not exist yet.  
From there it should be relativey easy to wrap it for users.  

The missing info at this stage is what constrants exactly there is on a file object in the 'JupyterLite Storage' IndexedDB.

```js
import os
import datetime as dt

import js, asyncio
from js import Object
from pyodide import to_js

DB_NAME = 'JupyterLite Storage'

async def get_contents(path):
    """use the IndexedDB API to acess JupyterLite's in-browser (for now) storage
    
    for documentation purposes, the full names of the JS API objects are used.
    
    see https://developer.mozilla.org/en-US/docs/Web/API/IDBRequest
    """
    # we only ever expect one result, either an error _or_ success
    queue = asyncio.Queue(1)
    
    IDBOpenDBRequest = js.self.indexedDB.open(DB_NAME)
    IDBOpenDBRequest.onsuccess = IDBOpenDBRequest.onerror = queue.put_nowait
    
    await queue.get()
    
    if IDBOpenDBRequest.result is None:
        return None
        
    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readonly")
    IDBObjectStore = IDBTransaction.objectStore("files")
    IDBRequest = IDBObjectStore.get(path, "key")
    IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait
    
    await queue.get()
    
    return IDBRequest.result.to_py() if IDBRequest.result else None



async def put_contents(content, path, overwrite=False):
    """
    """
    # count existing
    queue = asyncio.Queue(1)
    
    IDBOpenDBRequest = js.self.indexedDB.open(DB_NAME)
    IDBOpenDBRequest.onsuccess = IDBOpenDBRequest.onerror = queue.put_nowait
    await queue.get()
    
    if IDBOpenDBRequest.result is None:
        return None
        
    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readonly")
    IDBObjectStore = IDBTransaction.objectStore("files")
    IDBRequest = IDBObjectStore.count(path)
    
    IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait
    await queue.get()
    
    count = IDBRequest.result
    # print(f'count = {count}')
    
    if count == 1 and not overwrite:
        print(f'file {path} exists - will not overwrite')
        return 
    
    # add file
    value = {
        'name': os.path.basename(path), 
        'path': path,
        'format': 'text',
        'created': dt.datetime.now().isoformat(),
        'last_modified': dt.datetime.now().isoformat(),
        'content': content,
        'mimetype': 'text/plain',
        'type': 'file',
        'writable': True,
    }
    print(value)

    IDBTransaction = IDBOpenDBRequest.result.transaction("files", "readwrite")
    IDBObjectStore = IDBTransaction.objectStore("files")
    # see https://github.com/pyodide/pyodide/issues/1529#issuecomment-905819520
    value_as_js_obj = to_js(value, dict_converter=Object.fromEntries)
    if count == 0:
        IDBRequest = IDBObjectStore.add(value_as_js_obj, path)
    if count == 1:
        IDBRequest = IDBObjectStore.put(value_as_js_obj, path)
    IDBRequest.oncomplete = IDBRequest.onsuccess = IDBRequest.onerror = queue.put_nowait
    await queue.get()
    
    return IDBRequest.result
```

Then use it in cells:

+ read file:
```js
data_aa = await get_contents("data/aa.txt")
data_aa
# output:
{'name': 'aa.txt',
 'path': 'data/aa.txt',
 'last_modified': '2022-03-25T23:43:24.180Z',
 'created': '2022-03-25T23:43:07.595Z',
 'format': 'text',
 'mimetype': 'text/plain',
 'content': 'this is file aa.txt\n',
 'size': None,
 'writable': True,
 'type': 'file'}
```

+ write file:
```js
content = 'coucou i am a new file\nyeah! write works!'
path = "data/bb.txt"
r = await put_contents(content, path, overwrite=False)
r
# output
{'name': 'bb.txt', 'path': 'data/bb.txt', 'format': 'text', 'created': '2022-03-26T01:08:49.912000', 'last_modified': '2022-03-26T01:08:49.913000', 'content': 'coucou i am a new file\nyeah! it works!', 'mimetype': 'text/plain', 'type': 'file', 'writable': True}
data/bb.txt
```
+ overwrite file:
```js
content = 'coucou i am a new file\nyeah! overwrite works too!'
path = "data/bb.txt"
r = await put_contents(content, path, overwrite=True)
r
# output
{'name': 'bb.txt', 'path': 'data/bb.txt', 'format': 'text', 'created': '2022-03-26T01:09:36.392000', 'last_modified': '2022-03-26T01:09:36.392000', 'content': 'coucou i am a new file\nyeah! overwrite works too!', 'mimetype': 'text/plain', 'type': 'file', 'writable': True}
data/bb.txt
```

**Reply by bollwyvl** [reply]
Right, this would work, _today_, but would not work with an alternate storage backend e.g. #315.

Along the way, to support multiple deployments on the same domain, we also made the storage name configurable. 

Ideally, all emscripten-based systems would get this "for free" by mounting the contents with as a custom FS, such as [this one by starboard](https://github.com/gzuidhof/starboard-python/blob/88327bb56f7e2f673cf1ea4c5ff1e0a68c50e288/src/worker/emscripten-fs.ts#L120)... there are a number of angles to consider, as we'd want to make sure that it works transparently across webworkers, iframes, and other encapsulation techniques.

**Reply by psychemedia** [reply]
@oscar6echo

Trying the read function on the demo JupyterLite against, eg `data/iris.csv` file doesn't seem to work? I noticed in passing that @bollywvl's recipe for reading files seemed to break a couple or more weeks ago (not sure which update did it). Until such a time as a file read example appears in a tested notebook, I'd be wary of trusting it unless you use it against a pinned version of JupyterLite you know it works with.

**Reply by oscar6echo** [reply]
Indeed it does not work as the data "embeded" in a JLite instance does not exist in the "JupyterLite Storage"  indexedDB (IDB), unless/until you modify it then it is added to the IDB.

To access an embedded file e.g. simple fetch it from /files/data/iris.csv e.g.

```py
from js import fetch
from io import StringIO
import pandas as pd

res = await fetch('http://localhost:8081/files/data/iris.csv')
text = await res.text()
df = pd.read_csv(StringIO(text))
df.info()
```

To access a modified embeded file, then read it from the IDB.
To create then read new files, use the IDB.

To make sense of this complicated mix, I really advise to keep the brower console tab Application/IndexedDB/JupyterLite Storage/files - at least that's what I do to try and understand.

The benefit of this transparent access to IDB is that you can have some JS that lives entirely outside JLite writing data to it (and reacting to changes in it), you can then read it from the IDB and use it in your kernel. This is **VERY** convenient, I think.

**Reply by psychemedia** [reply]
@oscar6echo You can also bring a data file into browser storage by eg clicking on it in the JupyterLab file browser *and then saving it*.

But for a novice, this is really difficult to make sense of. The JupyterLab file browser is meaningless when it comes to being able to programmatically access a file, which at the moment can only (?) be done from browser storage? As a user, in JupyterLab, I can see the file, I can manually open it, but I can't see it from the py environment *unless* it has somehow previously found its way into browser storage.

Looking at the file browser does not tell you what state the file is in in terms of how you can access it. As you suggest, the only easy way you can tell by looking is to also open Dev tools and and inspect browser storage:

<img width="754" alt="image" src="https://user-images.githubusercontent.com/82988/160394885-dbd3a03f-581a-4b27-a1c0-97f48d0a5b1a.png">

What we really need is either a UI tweak to show what files are or are not in memory, eg use italics to show files are not in Browser Storage / available to code, or bold font to show they are. Or code that will let us push files that are listed in the file browser into browser storage either manually, from code, at the click of a UI button, or automatically behind the scenes (each of which cases may incur their own issues...)

At the moment, the safest way to distribute data files in a JupyterLite context seems to be distribute an environment that does not contain any data files at all, but to force the user to load them in from a remote URL and then save them using eg your `put_contents()` function. This will have the effect of displaying them in the file browser (which is meaningless as far as knowing whether you can access the file from code) and as a side effect put it into browser storage from where the `get_contents()` function can read them.

**Reply by oscar6echo** [reply]
I agree embedding/uploading/reading/writing files is a bit confusing a the moment, but as another regular user, I think JupyterLite has *massive* potential, dwarfing the current (probably temporary) convoluted ways of the "file system & co". 

> to force the user to load them in from a remote URL

No need. All embedded files are available on local url /files e.g. http://localhost:8081/files/data/iris.csv

> Looking at the file browser does not tell you what state the file is in in terms of how you can access it.

True but if you want to programatically read a file, just check if available in IDB, if not fetch from /files then write to IDB. Then all files needed by the kernel are in IDB.  
Of course you would want to wrap this behavior in a `jlite_utils` package, and add it in `/py` folder at build time. While this is tedious to do for an end user, it is quite easy to build and distribute to a small audience - small because all this is probably temporary.

Remark: For a short `requests` erzatz you can leverage `pyfetch` from `pyodide.http`:

```py
from pyodide.http import pyfetch

async def lite_request(
    url, 
    method='GET',
    headers=None, 
    params=None, 
    body=None
):
    """
    """
    if method == 'GET' and params is not None:
        suffix = '&'.join([f'{k}={v}' for k, v in params.items()])
        url += '?'+suffix
    
    if method == 'POST' and body is not None:
        body = json.dumps(body)
    
    headers = await build_auth_headers()
    
    res = await pyfetch(
        url, 
        method=method, 
        headers=headers,
        body=body,
    )    
    txt = await res.string()
    data = json.loads(txt)
    return data

# example
url = 'https://reqres.in/api/users'
method = 'GET'
params = {'page': 2}
await lite_request(url, params=params)

# example
url = 'https://reqres.in/api/users'
method = 'POST'
body = {"name": "morpheus","job": "leader"}
await lite_request(url, method=method, body=body)

```

NOTE: I said "very **CONVENIENT**" in my previous comment, I should say "very **POWERFUL**".

**Reply by psychemedia** [reply]
Re: `All embedded files are available on local url /files e.g. http://localhost:8081/files/data/iris.csv` But as others have pointed out elsewhere, programmatically, how do you reliably get access to the domain and path (`http://localhost:8081/` etc.)? There is no in-built supported way of introspecting it.

**Reply by psychemedia** [reply]
Re: "as another regular user" :-) I am always up front with my own peculiar user requirements: open distance education with very large, largely unsupported, populations of naive, non-technical, first time users, on a wide range of BYOD platforms. If there is any ambiguity/confusion, it makes things a non-starter for a lot of people, and therefore not a viable solution for me... Where things are capable of running in different places (eg Jupyter notebooks) they also have to run the same everywhere. If a notebook has lots of boilerplate logic to call things differently for eg jupyterlite or colab, it's a non-starter, not least because it just adds confusion and detracts from the point (the content). The more requirements in the notebook, the less portable. And if things are installed by magic behind the scenes, when learners try things themselves and they don't work because they didn't install something that was silently installed somewhere else, that's a non-starter too! ;-)

**Reply by oscar6echo** [reply]
To determine protocol, host, port you can do:

```py
import js

f = js.self.location.pathname
a, b = f.split('//')
c, _ = b.split('/')
protocol = a[:-1]
hostname, port = c.split(':')

print(protocol)
print(hostname)
print(port)
```

Re your more general point, I would say JLite in its current form ticks *some* boxes but not all:
+ web based so large audience
+ no config as any website
+ super scalable as user machine does all the work - no cloud resource - just CDN - quasi zero cost
+ allow people to focus on the actual python code in the *exact* setup established by the builder - who need be a bit technical at this stage  

But:
+ code is not directly portable to/from "Desktop Jupyter": a full JLite notebook will not generally work on Desktop Jupyter but pure Python snippets outside the "import" et "retrieve data" cells will.

So if the goal is to "teach notebooks" and how to become autonomous with them, probably not the best option. But to make a "specialized" collection of notebooks  available for consumption, it is valid, I would say. But it requires some investment to understand and workaround the unintuitive behavior.

In my case, I use it to provide a client on a protected API with some relatively flexible postprocessing - a swagger/graphiql companion if you like.

**Reply by psychemedia** [reply]
I seem to recall @jtpio saying that the `js.location` thing was not reliable. (As per my previous, I am personally quite happy working with things that only work half the time, but for end-users, they have to work reliably...)

I think there there is great potential around jupyterlite, particular for novices (install free etc), but it's very problematic w/ that user group because there are some very peculiar quirks and workarounds required for basic operations (eg file read/write, file deletion/reset to original distributed file state) and (as a practical class management issue), a lack of synch to desktop filesystem.

Also, despite my comments, I am a *huge* fan of jupyterlite; I just think it could disappoint or confuse a lot of people who want to use it to "just get started in an install free way" and then find eg their users can't open files, or lose files or etc etc. There are workarounds, but novices are fragile creatures... ;-)

**Reply by psychemedia** [reply]
FWIW, my collection of tricks is here: https://github.com/innovationOUtside/ouseful_jupyterlite_utils Is it okay if I add your `put_contents()` function?

**Reply by oscar6echo** [reply]
Sure ! Pick whatever you wish.   
It needs fine tuning though, for ex in case of update keep `created` unchanged, `mime/type` not always `text/plain`.

**Reply by oscar6echo** [reply]
NOTE: With the release of Version [0.1.0-beta.9](https://jupyterlite.readthedocs.io/en/latest/changelog.html), in particular enhancement "**Implement a custom Emscripten File System which communicates with the JupyterLab Content Manager, giving file access to pyolite** [#655](https://github.com/jupyterlite/jupyterlite/pull/655)", the hoop-jumping above is not longer necessary.  

Super ! :clap: 

````py
import os

os.getcwd()
# /drive

# to list files & folders in contents shown in file browser
os.listdir('/drive')
````

**Comment by jtpio** [comment, marked as answer]
Posting here for visibility.

`0.1.0b9` was released with support for file access (added in https://github.com/jupyterlite/jupyterlite/pull/655).

![image](https://user-images.githubusercontent.com/591645/174678441-a3e1af52-e681-4cf0-83c2-061a48341d13.png)
